src/v/rpc/test/rpc_gen_cycling_test.cc

```diff
@@ -73,6 +78,123 @@ FIXTURE_TEST(rpcgen_tls_integration, rpc_integration_fixture) {
+    // Server starts with bad credentials, files are updated on disk and then
+    // client connects. Expected behavior is that client can connect without
+    // issues. Condition variable is used to wait for credentials to reload.
+    temporary_dir tmp;
+    // client credentials
+    auto client_key = tmp.copy_file("redpanda.key", "client.key");
+    auto client_crt = tmp.copy_file("redpanda.crt", "client.crt");
+    auto client_ca = tmp.copy_file(
+      "root_certificate_authority.chain_cert", "ca_client.pem");
+    auto client_creds_builder = config::tls_config(
+                                  true,
+                                  config::key_cert{
+                                    client_key.native(), client_crt.native()},
+                                  client_ca.native(),
+                                  true)
+                                  .get_credentials_builder()
+                                  .get0();
+    // server credentials
+    auto server_key = tmp.copy_file("redpanda.other.key", "server.key");
+    auto server_crt = tmp.copy_file("redpanda.other.crt", "server.crt");
+    auto server_ca = tmp.copy_file(
+      "root_certificate_authority.other.chain_cert", "ca_server.pem");
+    auto server_creds_builder = config::tls_config(
+                                  true,
+                                  config::key_cert{
+                                    server_key.native(), server_crt.native()},
+                                  server_ca.native(),
+                                  true)
+                                  .get_credentials_builder()
+                                  .get0();
+
+    std::unordered_set<ss::sstring> updated;
+    ss::condition_variable cvar;
+    configure_server(
+      server_creds_builder,
+      [&updated, &cvar](
+        const std::unordered_set<ss::sstring>& delta,
+        std::exception_ptr const& err) {
+          info("server credentials reload event");
+          if (err) {
+              try {
+                  std::rethrow_exception(err);
+              } catch (...) {
+                  // The expection is expected to be triggered when the
+                  // temporary files are deleted at the end of the test
+                  info(
+                    "tls reloadable credentials callback exception: {}",
+                    std::current_exception());
+              }
+          } else {
+              for (auto const& name : delta) {
+                  updated.insert(name);
+              }
+              cvar.signal();
+          }
+      });
+
+    register_services();
+    start_server();
+    info("server started");
+
+    // fix client credentials and reconnect
+    info("replacing files");
+    tmp.copy_file("redpanda.key", "server.key");
+    tmp.copy_file("redpanda.crt", "server.crt");
+    tmp.copy_file("root_certificate_authority.chain_cert", "ca_server.pem");
+
+    cvar.wait([&updated] { return updated.size() == 3; }).get();
```

Comment:
where does the test wait for the server/client to start up with the incompatible certs? is it in `start_server`? that would be important to have so that the test wasn't racing in a way such that the certs were replaced before the server started up and the reload code paths weren't exercerised.

Reply:
oh, i see. that callback passed to configure_server is actually invoked specifically when the cert reload happens?